core/src/main/java/tech/pegasys/web3signer/core/service/http/handlers/keymanager/imports/ImportKeystoresHandler.java

```diff
@@ -0,0 +1,257 @@
+import io.vertx.ext.web.api.RequestParameters;
+import org.apache.commons.lang.StringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+
+public class ImportKeystoresHandler implements Handler<RoutingContext> {
+
+  private static final Logger LOG = LogManager.getLogger();
+  public static final int SUCCESS = 200;
+  public static final int BAD_REQUEST = 400;
+  public static final int SERVER_ERROR = 500;
+  private static final ObjectMapper YAML_OBJECT_MAPPER = new ObjectMapper(new YAMLFactory());
+
+  private final ObjectMapper objectMapper;
+  private final Path keystorePath;
+  private final Optional<SlashingProtection> slashingProtection;
+  private final ArtifactSignerProvider artifactSignerProvider;
+
+  public ImportKeystoresHandler(
+      final ObjectMapper objectMapper,
+      final Path keystorePath,
+      final Optional<SlashingProtection> slashingProtection,
+      final ArtifactSignerProvider artifactSignerProvider) {
+    this.objectMapper = objectMapper;
+    this.keystorePath = keystorePath;
+    this.slashingProtection = slashingProtection;
+    this.artifactSignerProvider = artifactSignerProvider;
+  }
+
+  @Override
+  public void handle(RoutingContext context) {
+    // API spec - https://github.com/ethereum/keymanager-APIs/tree/master/flows#import
+    final RequestParameters params = context.get("parsedParameters");
+    final ImportKeystoresRequestBody parsedBody;
+    try {
+      parsedBody = parseRequestBody(params);
+    } catch (final IllegalArgumentException | JsonProcessingException e) {
+      handleInvalidRequest(context, e);
+      return;
+    }
+
+    // check that keystores have matching passwords
+    if (parsedBody.getKeystores().size() != parsedBody.getPasswords().size()) {
+      context.fail(BAD_REQUEST);
+      return;
+    }
+
+    // no keystores passed in, nothing to do, return 200 and empty response.
+    if (parsedBody.getKeystores().isEmpty()) {
+      try {
+        context
+            .response()
+            .putHeader(CONTENT_TYPE, JSON_UTF_8)
+            .setStatusCode(SUCCESS)
+            .end(
+                objectMapper.writeValueAsString(
+                    new ImportKeystoresResponse(Collections.emptyList())));
+      } catch (JsonProcessingException e) {
+        context.fail(SERVER_ERROR, e);
+      }
+      return;
+    }
+
+    // extract pubkeys to import first
+    final List<String> pubkeysToImport = new ArrayList<>();
+    try {
+      pubkeysToImport.addAll(
+          parsedBody.getKeystores().stream()
+              .map(json -> new JsonObject(json).getString("pubkey"))
+              .map(
+                  key ->
+                      key.startsWith("0x")
+                          ? key
+                          : "0x" + key) // always use 0x prefix for comparisons
+              .collect(Collectors.toList()));
```

Comment:
does it make sense to convert incoming pubkey to lowercase and then append 0x if required? Otherwise do we have a testcase with mixed case public keys?

Reply:
I can lowercase everything for safety. What's the best place to put a utility method to normalize keys?

Reply:
nevermind - looks like we already have `IdentifierUtils.normaliseIdentifier()`

I'll just use that

Reply:
done - cc @jframe